src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,615 @@
+/*
+ * Copyright 2022-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mc-fle2-rfds-private.h"
+
+#include "mongocrypt-private.h"          // CLIENT_ERR
+#include "mc-fle-blob-subtype-private.h" // MC_SUBTYPE_FLE2EncryptionPlaceholder
+
+#include <math.h>        // INFINITY
+#include "mlib/thread.h" // mlib_once_flag
+
+static mc_FLE2RangeOperator_t
+get_operator_type (const char *key)
+{
+   BSON_ASSERT_PARAM (key);
+
+   if (0 == strcmp (key, "$gt")) {
+      return FLE2RangeOperator_kGt;
+   } else if (0 == strcmp (key, "$gte")) {
+      return FLE2RangeOperator_kGte;
+   } else if (0 == strcmp (key, "$lt")) {
+      return FLE2RangeOperator_kLt;
+   } else if (0 == strcmp (key, "$lte")) {
+      return FLE2RangeOperator_kLte;
+   } else {
+      return FLE2RangeOperator_kNone;
+   }
+}
+
+static const char *
+mc_FLE2RangeOperator_to_string (mc_FLE2RangeOperator_t op)
+{
+   switch (op) {
+   case FLE2RangeOperator_kGt:
+      return "$gt";
+   case FLE2RangeOperator_kGte:
+      return "$gte";
+   case FLE2RangeOperator_kLt:
+      return "$lt";
+   case FLE2RangeOperator_kLte:
+      return "$lte";
+   case FLE2RangeOperator_kNone:
+      return "none";
+   default:
+      return "Unknown";
+   }
+   return "Unknown";
+}
+
+static bool
+is_supported_operator (const char *key)
+{
+   BSON_ASSERT_PARAM (key);
+   return get_operator_type (key) != FLE2RangeOperator_kNone;
+}
+
+// Suffix an error message with the JSON string form of bson.
+#define ERR_WITH_BSON(bson, fmt, ...)                            \
+   if (1) {                                                      \
+      char *_str = bson_as_canonical_extended_json (bson, NULL); \
+      CLIENT_ERR (fmt ": %s", __VA_ARGS__, _str);                \
+      bson_free (_str);                                          \
+   } else                                                        \
+      (void) 0
```

Comment:
```suggestion
#define ERR_WITH_BSON(bson, fmt, ...)                              \
   if (1) {                                                        \
      char *_str = bson_as_canonical_extended_json ((bson), NULL); \
      CLIENT_ERR (fmt ": %s", __VA_ARGS__, _str);                  \
      bson_free (_str);                                            \
   } else                                                          \
      (void) 0
```
Guard expansion of `bson`.